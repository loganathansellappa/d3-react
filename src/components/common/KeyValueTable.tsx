import React from "react";

interface KeyValueTableProps {
  data: { [key: string]: string }[];
  headers: Array<string>;
}

export const KeyValueTable: React.FC<KeyValueTableProps> = ({
  data,
  headers,
}) => {
  // Extract all unique keys from the data
  const allKeys = Array.from(
    new Set(data.flatMap((item) => Object.keys(item))),
  );

  return (
    <table>
      <thead>
        <tr>
          {headers.map((h, index) => (
            <th key={index}>{h}</th>
          ))}
        </tr>
      </thead>
      <tbody>
        {allKeys.map((key) => (
          <tr key={key}>
            <td>{key}</td>
            {data.map((item, itemIndex) => (
              <td key={itemIndex}>{item[key]}</td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
};
