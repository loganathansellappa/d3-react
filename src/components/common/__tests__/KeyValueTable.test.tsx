import { render } from "@testing-library/react";
import { KeyValueTable } from "../KeyValueTable";

describe("KeyValueTable", () => {
  let element: { container: HTMLElement };
  beforeEach(() => {
    const props = {
      data: [
        { name: "NameOne", value: "ValueOne" },
        { name: "NameTwo", value: "ValueTwo" },
      ],
      headers: ["Name", "Value"],
    };
    element = render(<KeyValueTable {...props} />);
  });

  test("renders KeyValueTable component", () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });

  test("displays tables with data", () => {
    expect(element.container.textContent).toContain("NameOne");
    expect(element.container.textContent).toContain("NameTwo");
  });
});
