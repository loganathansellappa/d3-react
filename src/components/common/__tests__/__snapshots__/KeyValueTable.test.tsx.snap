// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyValueTable renders KeyValueTable component 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Value
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        name
      </td>
      <td>
        NameOne
      </td>
      <td>
        NameTwo
      </td>
    </tr>
    <tr>
      <td>
        value
      </td>
      <td>
        ValueOne
      </td>
      <td>
        ValueTwo
      </td>
    </tr>
  </tbody>
</table>
`;
